samples/vertex-ffp/main.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * This sample provides a very basic demonstration of 3D rendering on the Xbox,
+ * using pbkit. Based on the pbkit demo sources.
+ */
+#include <hal/video.h>
+#include <hal/xbox.h>
+#include <math.h>
+#include <pbkit/pbkit.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <xboxrt/debug.h>
+
+#define pb_end(x) 0; pb_end(x)
+
+#include <xgu/xgu.h>
+#include <xgu/xgux.h>
+
+#include "xgu_extra.h"
+#include "xgux_extra.h"
+
+static uint32_t *alloc_vertices;
+static uint32_t  num_vertices;
+static float     m_viewport[4][4];
+
+typedef struct {
+    float pos[3];
+    float color[3];
+} __attribute__((packed)) ColoredVertex;
+
+static const ColoredVertex verts[] = {
+    //  X     Y     Z       R     G     B
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}}, /* Background triangle 1 */
+    {{-1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}}, /* Background triangle 2 */
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0, -1.0,  1.0}, { 1.0,  0.0,  0.0}}, /* Foreground triangle */
+    {{ 0.0,  1.0,  1.0}, { 0.0,  1.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.0,  0.0,  1.0}},
+};
+
+/* Construct a viewport transformation matrix */
+static void matrix_viewport(float out[4][4], float x, float y, float width, float height, float z_min, float z_max)
+{
+    memset(out, 0, 4*4*sizeof(float));
+    out[0][0] = width/2.0f;
+    out[1][1] = height/-2.0f;
+    out[2][2] = z_max - z_min;
+    out[3][3] = 1.0f;
+    out[3][0] = x + width/2.0f;
+    out[3][1] = y + height/2.0f;
+    out[3][2] = z_min;
+}
+
+/* Main program function */
+int main(void)
+{
+    uint32_t *p;
+    int       start, last, now;
+    int       fps, frames, frames_total;
+
+    /* Set display mode */
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    /* Start GPU */
+    int status = pb_init();
+    if (status) {
+        debugPrint("pb_init Error %d\n", status);
+        Sleep(2000);
+        return 1;
+    }
+
+    //FIXME: Remove.
+    //       This works around a race-condition in XQEMU.
+    //       It looks like pb_init uses `debugPrint()` so the GPU and CPU draw
+    //       at the same time. Depending on the order of operations, this
+    //       breaks the XQEMU surface cache.
+    Sleep(100);
+
+    /* Display GPU result on display */
+    pb_show_front_screen();
+
+    /* Basic setup */
+    int width = pb_back_buffer_width();
+    int height = pb_back_buffer_height();
+
+    /* Setup fragment shader */
+    p = pb_begin();
+    #include "ps.inl"
+    pb_end(p);
+
+    /* Load constant geometry */
+    alloc_vertices = MmAllocateContiguousMemoryEx(sizeof(verts), 0, 0x3ffb000, 0, 0x404);
```

Comment:
This allocator is pretty bad as it has magic numbers. It came from the triangle sample (as much of the rest)